edb/pgsql/compiler/relctx.py

```diff
@@ -1413,20 +1417,15 @@ def range_for_material_objtype(
     include_overlays: bool=True,
     include_descendants: bool=True,
     ignore_rewrites: bool=False,
+    is_global: bool=False,
     dml_source: Sequence[irast.MutatingLikeStmt]=(),
     ctx: context.CompilerContextLevel,
 ) -> pgast.PathRangeVar:
 
     env = ctx.env
 
-    # If this is a view type, but it still appears in rewrites, that is
-    # because it is a global that we are caching.
-    if (
-        typeref.material_type is not None
-        and (typeref.id, include_descendants) not in ctx.env.type_rewrites
-    ):
-        typeref = typeref.material_type
```

Comment:
This doesn't work because the shape on the global causes a new derived type, which will not appear in `type_rewrites`, right?

Reply:
It was sillier than that, I think. When there was a shape, the derived type appeared *inside* the `type_rewrite` also, and so it got picked up as being a global (since it *was* in type_rewrites), but it didn't get compiled that way because we were already compiling it.
Instead of trying to tease out that case, I made the mechanism more explicit.